services/data/src/links/RestAPILink/fetchData.ts

```diff
@@ -10,33 +10,66 @@ export const parseContentType = (contentType: string | null) => {
 }
 
 export const parseStatus = async (response: Response) => {
-    if (
+    const accessError =
         response.status === 401 ||
         response.status === 403 ||
+        response.status === 404 ||
         response.status === 409
-    ) {
-        const message = await response
-            .json()
-            .then(body => {
-                return body.message
-            })
-            .catch(() => {
-                return response.status === 401 ? 'Unauthorized' : 'Forbidden'
-            })
+
+    if (accessError) {
+        let message
+        let details = {} as any
+
+        try {
+            details = await response.json()
+
+            if (typeof details === 'string') {
+                message = details
+
+                // Clearing details, since it's expected to be an object and not a string
+                details = {}
+            } else if (typeof details === 'object') {
+                message = details.message
+            }
```

Comment:
So I did a quick test in Chrome console and it looks like the "raw text" option here is a bit more nuanced:

```js
> (new Response("This is a test")).json().then(console.log)
VM1441:1 Uncaught (in promise) SyntaxError: Unexpected token T in JSON at position 0
> (new Response('"This is a test"')).json().then(console.log)
This is a test
```

So I think the "string" case we're catching here (wrapped in double quotes) will actually be super rare.

I think we should actually keep this as-is, just try to parse the body as JSON, assume it's an object, and use `details.message` if it exists.  So something like this:

```js
try {
    details = await response.json()
    message = details.message
} catch (e) {
    // Do nothing
}
// Moving this outside the catch means we put always have something in `message` even if the returned JSON doesn't have a `message` property
if (!message) {
    message = response.status === 401 ? 'Unauthorized' : 'Forbidden'
}
```

---
I initially took a stab at a more "exhaustive" option, which will also accept non-JSON bodies and use make a custom details object out of them.  I'm not sure I like it as much... it is liberal in what it accepts from the server, but maybe in this case since we control the server we should be more conservative and fix the unexpected responses before the server sends them?

```js
const errorText = await response.text()
try {
    details = await JSON.parse(errorText) // I think we can assume this is an object here...
    message = details.message
} catch (e) {
    message = errorText
}
if (!message || !message.length) {
    message = response.status === 401 ? 'Unauthorized' : 'Forbidden'
}
```

Let me know which you prefer!

Reply:
I'd prefer the former. I think it's better to assume uniformity across our server responses. If we don't have that, it'd be good to have it fail in my opinion instead of smoothing it over.